### DISCUSSION TITLE
Atomic Insert feature request

### DISCUSSION THREAD

**Discussion by koculu** [microsoft/FASTER#673] [Ideas]
It is possible to implement atomic inserts using RMW but it is not easy to use and is not most efficient.
Would it be possible to implement an Insert method which does insert only if there isn't any.
Method can be: 
bool TryInsert(key, value)

**Comment by badrishc** [comment]
The RMW operation is designed exactly for this class of conditional updates, and it should indeed be optimal in performance, i.e., no work is done that could have been avoided with a more direct `Insert` API. Here is the approximate recipe for the `Insert` operation:

NeedInitialUpdate => true
InitialUpdater => write input to value
NeedCopyUpdate => false
CopyUpdater => not implemented
InPlaceUpdater => return true

**Reply by koculu** [reply]
InPlaceUpdate => shouldn't return false in the TryInsert scenario?

**Reply by badrishc** [reply]
On IPU returns false, FASTER will assume there was no space for IPU and therefore allocate space on tail of log and perform a CopyUpdate. Returning true tells FASTER that the IPU succeeded so it can end the operation, even though we did not actually perform a write on the value.
